modules/renter/download.go

```diff
@@ -1,805 +1,484 @@
-}
+	// Add the download object to the download queue.
+	r.downloadHistoryMu.Lock()
+	r.downloadHistory = append(r.downloadHistory, d)
+	r.downloadHistoryMu.Unlock()
 
-// DownloadHTTPWriter is a http response writer-backed implementation of
-// DownloadWriter.  The writer writes all content that is written to the
-// current `offset` directly to the ResponseWriter, and buffers all content
-// that is written at other offsets.  After every write to the ResponseWriter
-// the `offset` and `length` fields are updated, and buffer content written
-// until
-type DownloadHTTPWriter struct {
-	w              io.Writer
-	offset         int            // The index in the original file of the last byte written to the response writer.
-	firstByteIndex int            // The index of the first byte in the original file.
-	length         int            // The total size of the slice to be written.
-	buffer         map[int][]byte // Buffer used for storing the chunks until download finished.
-}
-
-// NewDownloadHTTPWriter creates a new instance of http.ResponseWriter backed DownloadWriter.
-func NewDownloadHTTPWriter(w io.Writer, offset, length uint64) *DownloadHTTPWriter {
-	return &DownloadHTTPWriter{
-		w:              w,
-		offset:         0,           // Current offset in the output file.
-		firstByteIndex: int(offset), // Index of first byte in original file.
-		length:         int(length),
-		buffer:         make(map[int][]byte),
+	// Block until the download has completed.
+	select {
+	case <-d.completeChan:
+		return d.Err()
+	case <-r.tg.StopChan():
+		return errors.New("download interrupted by shutdown")
 	}
 }
 
-// Destination implements the Destination method of the DownloadWriter
-// interface and informs callers where this download writer is
-// being written to.
-func (dw *DownloadHTTPWriter) Destination() string {
-	return "httpresp"
-}
-
-// Close implements DownloadWriter's Close method.
-func (dw *DownloadHTTPWriter) Close() error {
-	return nil
-}
-
-// WriteAt buffers parts of the file until the entire file can be
-// flushed to the client. Returns the number of bytes written or an error.
-func (dw *DownloadHTTPWriter) WriteAt(b []byte, off int64) (int, error) {
-	// Write bytes to buffer.
-	offsetInBuffer := int(off) - dw.firstByteIndex
-	dw.buffer[offsetInBuffer] = b
-
-	// Send all chunks to the client that can be sent.
-	totalDataSent := 0
-	for {
-		data, exists := dw.buffer[dw.offset]
-		if exists {
-			// Send data to client.
-			dw.w.Write(data)
-
-			// Remove chunk from map.
-			delete(dw.buffer, dw.offset)
-
-			// Increment offset to point to the beginning of the next chunk.
-			dw.offset += len(data)
-			totalDataSent += len(data)
+// DownloadHistory returns the list of downloads that have been performed. Will
+// include downloads that have not yet completed. Downloads will be roughly, but
+// not precisely, sorted according to start time.
+//
+// TODO: Currently only remembers downloads from the current session.
```

Comment:
I think that people will start looking at these TODOs at some point. Once you've gotten started with some issues, gotten familiar with the codebase, then it makes sense to dive into the code and start tackling all the TODOs lying around.

Also, no reason it can't both be a todo and a programming task.

Perhaps we can add a programming task to take all of the TODOs and turn them into programming tasks :)